### PR TITLE
[DynamicShape] Add dynamic shape support for builder::mean 

### PR DESCRIPTION
**Pull request by mbencer**

### CODE REVIEW

File: src/ngraph/builder/reduce_ops.cpp [hunk 1]
```diff
@@ -116,18 +129,33 @@ namespace ngraph
...
-            const auto& et = value.get_element_type();
-
-            const auto divisor = ngraph::opset1::Constant::create(et, Shape{}, {N});
+            if (value.get_partial_shape().is_static())
+            {
+                value_elems_sum = std::make_shared<ngraph::opset1::ReduceSum>(
+                    value,
+                    ngraph::opset1::Constant::create(
+                        element::i64, Shape{reduction_axes.size()}, reduction_axes.to_vector()),
+                    keep_dims);
```
Comment (arogowie-intel) [inline src/ngraph/builder/reduce_ops.cpp:142]: You can get this part out of this `if`.

File: src/ngraph/builder/reduce_ops.cpp [hunk 2]
```diff
@@ -116,18 +129,33 @@ namespace ngraph
...
+                auto elems_number_value = get_num_elements(value.get_shape(), reduction_axes);
+                elems_number = ngraph::opset1::Constant::create(
+                    value_elem_type, Shape{}, {elems_number_value});
+            }
+            else
+            {
+                const auto reduction_axes_const = ngraph::opset1::Constant::create(
+                    element::i64, Shape{reduction_axes.size()}, reduction_axes.to_vector());
+                value_elems_sum = std::make_shared<ngraph::opset1::ReduceSum>(
+                    value, reduction_axes_const, keep_dims);
```
Comment (arogowie-intel) [inline src/ngraph/builder/reduce_ops.cpp:152]: As above.

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -315,6 +315,7 @@ set(MULTI_TEST_SRC
     backend/batch_norm.in.cpp
     backend/broadcast.in.cpp
     backend/builder_flatten.in.cpp
+    backend/builder_opset1.in.cpp
```
Comment (arogowie-intel) [inline test/CMakeLists.txt:318]: Maybe renaming it `builder_reduce_ops_opset1.in.cpp`  would better fit to naming convention as file name in line above?

File: test/backend/builder_opset1.in.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "gtest/gtest.h"
+#include "ngraph/ngraph.hpp"
+#include "util/all_close_f.hpp"
+#include "util/autodiff/numeric_compare.hpp"
+#include "util/test_control.hpp"
+#include "util/test_tools.hpp"
```
Comment (arogowie-intel) [inline test/backend/builder_opset1.in.cpp:22]: Please clean unnecessary headers.

File: test/backend/builder_opset1.in.cpp [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+    const Shape expected_output_shape{4};
+    auto result = backend->create_tensor(element::f32, expected_output_shape);
+    handle->call_with_validate({result}, {input_tensor});
+
+    EXPECT_EQ(result->get_element_type(), input->get_element_type());
+    EXPECT_EQ(result->get_shape(), expected_output_shape);
+    const vector<float> expected_output{2.5f, 8.5f, 14.5f, 20.5f};
+    const auto output_values = read_vector<float>(result);
+    EXPECT_EQ(output_values, expected_output);
```
Comment (arogowie-intel) [inline test/backend/builder_opset1.in.cpp:52]: If you have created `Function` you may easily use `NgraphTestCase` class.

File: test/backend/builder_opset1.in.cpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+    auto result = backend->create_dynamic_tensor(element::f32, PartialShape::dynamic());
+    handle->call_with_validate({result}, {input_tensor});
+
+    EXPECT_EQ(result->get_element_type(), input->get_element_type());
+    const Shape expected_output_shape{5};
+    EXPECT_EQ(result->get_shape(), expected_output_shape);
+    const vector<float> expected_output{17.5f, 18.5f, 19.5f, 20.5f, 21.5f};
+    const auto output_values = read_vector<float>(result);
+    EXPECT_EQ(output_values, expected_output);
```
Comment (arogowie-intel) [inline test/backend/builder_opset1.in.cpp:79]: As above: you may use `NgraphTestCase`.

File: test/backend/builder_opset1.in.cpp [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+    auto result = backend->create_dynamic_tensor(element::f32, PartialShape::dynamic());
+    handle->call_with_validate({result}, {input_tensor});
+
+    EXPECT_EQ(result->get_element_type(), input->get_element_type());
+    const Shape expected_output_shape{2};
+    EXPECT_EQ(result->get_shape(), expected_output_shape);
+    const vector<float> expected_output{1.f, 4.f};
+    const auto output_values = read_vector<float>(result);
+    EXPECT_EQ(output_values, expected_output);
```
Comment (arogowie-intel) [inline test/backend/builder_opset1.in.cpp:107]: As above: you may use `NgraphTestCase`.

File: src/ngraph/builder/reduce_ops.cpp (hunk 1 above)
Comment (mbencer) [reply src/ngraph/builder/reduce_ops.cpp:142]: right, done

File: src/ngraph/builder/reduce_ops.cpp (hunk 2 above)
Comment (mbencer) [reply src/ngraph/builder/reduce_ops.cpp:152]: done

File: test/CMakeLists.txt (hunk 3 above)
Comment (mbencer) [reply test/CMakeLists.txt:318]: good point

File: test/backend/builder_opset1.in.cpp (hunk 4 above)
Comment (mbencer) [reply test/backend/builder_opset1.in.cpp:22]: done

File: test/backend/builder_opset1.in.cpp (hunk 5 above)
Comment (mbencer) [reply test/backend/builder_opset1.in.cpp:52]: done

File: test/backend/builder_opset1.in.cpp (hunk 6 above)
Comment (mbencer) [reply test/backend/builder_opset1.in.cpp:79]: done

File: test/backend/builder_opset1.in.cpp (hunk 7 above)
Comment (mbencer) [reply test/backend/builder_opset1.in.cpp:107]: done

**Comment by raramer01** [comment]
TESTNOW_DLDT_E2E

**Comment by raramer01** [comment]
TESTNOW_DLDT_E2E

**Pull request merged**